src/shared/html_or_plaintext.rs

```diff
@@ -102,6 +148,279 @@ live_design! {
+        if let ApplyFrom::NewFromDoc { .. } = apply.from {
+            let scope = apply.scope.as_ref().unwrap();
+            let doc = scope.props.get::<HtmlDoc>().unwrap();
+            let mut walker = doc.new_walker_with_index(scope.index + 1);
+            if let Some((live_id!(href), attr)) = walker.while_attr_lc() {
+                self.url = attr.into();
+            }
+        }
+    }
+}
+
+impl Widget for RobrixHtmlLink {
+    fn handle_event(&mut self, cx: &mut Cx, event: &Event, scope: &mut Scope) {
+        self.view.handle_event(cx, event, scope)
+    }
+
+    fn draw_walk(&mut self, cx: &mut Cx2d, scope: &mut Scope, walk: Walk) -> DrawStep {
+        let link = self.html_link(id!(link));
+        let matrix_link = self.matrix_link_pill(id!(matrix_link));
+
+        match self.match_link_type(&self.url) {
+            RobrixHtmlLinkType::Matrix => {
+                if let Some((matrix_id, via)) = self.handle_resolve_matrix_link(&self.url) {
+                    matrix_link.set_matrix_link_info(matrix_id, via);
+                    matrix_link.show(cx);
+                    matrix_link.draw_walk(cx, scope, walk)
+                } else {
+                    link.set_text(cx, &self.url);
+                    link.draw_walk(cx, scope, walk)
+                }
+            }
+            RobrixHtmlLinkType::Normal => {
+                link.set_text(cx, &self.url);
+                link.draw_walk(cx, scope, walk)
+            }
+        }
+    }
+
+    fn text(&self) -> String {
+        self.text.as_ref().to_string()
+    }
+
+    fn set_text(&mut self, cx:&mut Cx, v: &str) {
+        self.text.as_mut_empty().push_str(v);
+        self.redraw(cx);
+    }
+}
+
+impl RobrixHtmlLink {
+    fn match_link_type(&self, url: &str) -> RobrixHtmlLinkType {
+        if self.is_matrix_link(url) {
+            RobrixHtmlLinkType::Matrix
+        } else {
+            RobrixHtmlLinkType::Normal
+        }
+    }
+
+    fn is_matrix_link(&self, url: &str) -> bool {
+        MatrixToUri::parse(url).is_ok() || MatrixUri::parse(url).is_ok()
+    }
+
+    fn handle_resolve_matrix_link(&self, url: &str) -> Option<(MatrixId, Vec<OwnedServerName>)> {
+        MatrixToUri::parse(url)
+            .map(|uri| (uri.id().clone(), uri.via().to_vec()))
+            .or_else(|_| MatrixUri::parse(url).map(|uri| (uri.id().clone(), uri.via().to_vec())))
+            .ok()
+    }
+}
+
+#[derive(Live, LiveHook, Widget)]
+struct MatrixLinkPill {
+    #[deref] view: View,
+
+    #[rust] matrix_id: Option<MatrixId>,
+    #[rust] via: Vec<OwnedServerName>,
+
+    #[rust] name: String,
+    #[rust] avatar_url: Option<OwnedMxcUri>,
+}
+
+impl Widget for MatrixLinkPill {
+    fn handle_event(&mut self, cx: &mut Cx, event: &Event, scope: &mut Scope) {
+        let uid = self.widget_uid();
+        if self.visible {
+            if let Hit::FingerUp(fe) = event.hits(cx, self.area()) {
+                if fe.is_over {
```

Comment:
we need some more conditions here:
```suggestion
                if fe.is_over && fe.is_primary_hit() && fe.was_tap() {
```